File: app/src/main/java/com/cs/android/view/DividerItemDecoration.java
package com.cs.android.view;

import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.view.View;

import androidx.annotation.ColorInt;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

/**
 * @author ChenSen
 * @desc
 * @since 2021/12/21 13:18
 **/
public class DividerItemDecoration extends RecyclerView.ItemDecoration {

    private int mDividerHeight;
    private int mDividerColor;
    private int mLeftMargin;
    private int mRightMargin;
    private Paint mDividerPaint;

    public DividerItemDecoration(@ColorInt int dividerColor, int dividerHeight, int leftMargin, int rightMargin) {
        mDividerColor = dividerColor;
        mDividerHeight = dividerHeight;
        mLeftMargin = leftMargin;
        mRightMargin = rightMargin;

        mDividerPaint = new Paint();
        mDividerPaint.setColor(mDividerColor);
        mDividerPaint.setAntiAlias(true);
        mDividerPaint.setStyle(Paint.Style.FILL);
    }

    @Override
    public void getItemOffsets(@NonNull Rect outRect, @NonNull View view, @NonNull RecyclerView parent, @NonNull RecyclerView.State state) {
        super.getItemOffsets(outRect, view, parent, state);
        if (parent.getChildLayoutPosition(view) == 0) {
            outRect.top = mDividerHeight;
        } else {
            outRect.top = 0;
        }
        outRect.bottom = mDividerHeight;
    }

    @Override
    public void onDraw(@NonNull Canvas c, @NonNull RecyclerView parent, @NonNull RecyclerView.State state) {
        int childCount = parent.getChildCount();
        int left = parent.getPaddingLeft() + mLeftMargin;
        int right = parent.getWidth() - parent.getPaddingRight() - mRightMargin;

        for (int i = 0; i < childCount; i++) {
            View view = parent.getChildAt(i);
            float top = view.getBottom();
            float bottom = top + mDividerHeight;
            c.drawRect(left, top, right, bottom, mDividerPaint);

            if (i == 0) {
                c.drawRect(left, view.getTop() - mDividerHeight, right, view.getTop(), mDividerPaint);
            }
        }
    }
}

File: common/src/main/java/com/cs/common/widget/DividerItemDecoration.kt
package com.cs.common.widget

import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Rect
import android.view.View
import androidx.annotation.ColorInt
import androidx.recyclerview.widget.RecyclerView

/**
 * @author ChenSen
 * @since 2021/7/22 9:26
 * @desc
 */
class DividerItemDecoration(
    private val height: Int = 0,
    @ColorInt private val dividerColor: Int = Color.GRAY
) : RecyclerView.ItemDecoration() {

    private val mPaint: Paint by lazy {
        Paint().apply {
            isAntiAlias = true
            color = dividerColor
            style = Paint.Style.FILL
        }
    }


    override fun onDraw(c: Canvas, parent: RecyclerView, state: RecyclerView.State) {
        super.onDraw(c, parent, state)
    }


    override fun onDrawOver(c: Canvas, parent: RecyclerView, state: RecyclerView.State) {
        if (parent.layoutManager == null || parent.childCount <= 1) {
            return
        }

        (0 until parent.childCount).forEach {
            val child = parent.getChildAt(it)

            val left = child.left
            val top = child.bottom
            val right = left + child.measuredWidth
            val bottom = top + height
            c.drawRect(left.toFloat(), top.toFloat(), right.toFloat(), bottom.toFloat(), mPaint)
        }

    }


    override fun getItemOffsets(
        outRect: Rect,
        view: View,
        parent: RecyclerView,
        state: RecyclerView.State
    ) {
        if (parent.getChildAdapterPosition(view) == 0) {
            outRect.set(0, 0, 0, 0)
        } else {
            outRect.set(0, height, 0, 0)
        }
    }

}